### PR TITLE
Firebase adapter

### PR DESCRIPTION
**Pull request by kpsrinivas**
### Type of change
New feature

### Description
An adapter to work with firebase. A single firebase project can be used with many projects. We are working on a feature to deploy assets to firestore storage.

### CODE REVIEW

**Comment by bedeoverend** [comment]
Thanks for the PR, looks great!

Would you mind reverting your changes to the README? Can then add Firebase docs to the main [docs repo](https://github.com/simplajs/docs)

**Comment by madeleineostoja** [comment]
@bedeoverend do we need to finalise (and document, but at least finalise) adapter interface, before merging this? In case of breaking changes

**Comment by kpsrinivas** [comment]
@seaneking I am with you. I think its better to finalise adapter interface. Better finalise the architecture. Also working on deploy to firebase with functions and SEO. Might as well not need github to store data. We can store that json in realtime database. @bedeoverend will update the docs.

**Comment by kpsrinivas** [comment]
I did write an initial Readme - https://github.com/heresy/simpla Sorry that I changed the README. We might have to revert it back to original and append this to it. I will do that.

**Comment by kpsrinivas** [comment]
If I may also suggest, can we split this to different packages ? I can write the storage functionality and https(using their socket instead of the rest API) library for firebase. People can install saying @simpla/firebase-auth @simpla/netlify @simpla/firebase-hosting or @simpla/github-hosting

**Comment by bedeoverend** [comment]
Happy to wait until we’ve finalized adapter architecture, though I’m not too concerned about breaking changes while the adapters are about of Simpla’s core. On that point, once we do settle up on architecture, anyone can create a package outside of core. Unless we do a mono repo, @seaneking thoughts?

Also FWIW definitely worth thinking about storage adapter, not just authentication - agree it’d be great to add firebase storage!

**Comment by bedeoverend** [comment]
Also with Christmas and New Years probably  won’t have time to look at this until mid Jan

**Comment by madeleineostoja** [comment]
Just re: docs - sorry if it wasn't clear, but we're not using the README here for documentation (the snippet with Netlify is just an example to give people a glimpse of Simpla). We'll need a PR with a guide and API reference in Simpla's [docs](https://github.com/simplajs/docs). Use `SimplaNetlify` adapter docs as a template.

And yep as @bedeoverend touched on, we'll be soon opening up the storage interface as well (#77), at which point you could make a seperate Firebase adapter for that. Haven't settled on a naming convention yet, but I'd guess something like `FirebaseAuth` and `FirebaseStorage` for them.

File: .gitignore [hunk 1]
```diff
@@ -2,7 +2,7 @@
 node_modules
 
 # Built files
-adapters
+/adapters
```
Comment (seaneking) [inline .gitignore]: Why was this changed?

Comment (madeleineostoja) [inline .gitignore:5]: Why was this changed?

File: README.md [hunk 2]
```diff
@@ -117,3 +117,37 @@ Read the [Contributing guidelines](/CONTRIBUTING.md) for more information.
 ***
 
 [MIT](/LICENSE) © 2017
+
```
Comment (madeleineostoja) [inline README.md:120]: As mentioned in comments, this should be moved to documentation at [simplajs/docs](https://github.com/simplajs/docs)

File: src/adapters/firebase.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+export default class FirebaseAuth{
+  constructor(firebase){
+    this.firebase = firebase;
+  }
+
+  authenticate(){
+    var provider = new this.firebase.auth.GithubAuthProvider();
```
Comment (madeleineostoja) [inline src/adapters/firebase.js:7]: This repo is written in ES6 and transpiled to ES5 - code style should be consistent. Ie: `var` -> `let`/`const`, anon functions -> arrow functions, spacing between {...} tokens, etc.

**Review by madeleineostoja — COMMENTED** [review]
Just a few minor comments

File: .gitignore (hunk 1 above)
Comment (kpsrinivas) [reply .gitignore:5]: I think thats how git works. The built files are in the root which should be ignored. With just adapters, it will ignore adapters in the src folder too. So any changes to netlify will also not reflect. The files that exist before adding .gitignore will be there.

File: README.md (hunk 2 above)
Comment (kpsrinivas) [reply README.md:120]: Sure will do that.

File: .gitignore (hunk 1 above)
Comment (madeleineostoja) [reply .gitignore:5]: Ah gotcha, forgot about the adapters folder in src. Good catch!

**Comment by madeleineostoja** [comment]
> If I may also suggest, can we split this to different packages ? I can write the storage functionality and https(using their socket instead of the rest API) library for firebase

Missed this comment before - we did consider making all adapters separate packages, but in the end decided it was too much overhead for no real gain. Bundling adapters doesn't add any weight for the user, since they only pull in what they need, and makes maintenance much easier.

That said you're more than welcome to turn this into your own package rather than a PR against the core lib. At some point we'll probably put up a list somewhere of community adapters as well as the 'official' bundled ones.

**Comment by kpsrinivas** [comment]
Made this a package - https://github.com/heresy/simpla-firebase-auth . I guess we can close this PR. @seaneking

**Comment by madeleineostoja** [comment]
:+1: Let us know if you'd like to make it part of the core bundle, we'd be happy to help maintain it if it was.

**Comment by madeleineostoja** [comment]
Also feel free to make a PR against docs for your adapter (for discoverability), can make an official/community distinction there.

**Comment by kpsrinivas** [comment]
Its just one function and file. We can do as you please. I will write update the doc.
